### ISSUE TITLE
predict with multi-gpu and ddp strategy only get a portion of the predictions

### ISSUE THREAD

**Issue by gianscarpe** [PyTorchLightning/pytorch-lightning#10618]
## 🐛 Bug
When called 'predict' method in multi-gpu settings (e.g., 2 gpus, strategy "ddp"), I get only half the predictions (meaning that only the first half of the dataset returns the predictions)

### To Reproduce
I opened a PR with the failing test in 'test_gpu.py"
PR #10619 

### Expected behavior

I expect to get the full predictions for input dataset, potentially on CPU to avoid GPU conflicts
### Additional context
I believe torch.distributed has the right APIs to implement this (if this is really bug). Let me know! :)

cc @justusschock @kaushikb11 @awaelchli @akihironitta

**Comment by awaelchli** [comment]
If I recall correctly (@tchaton), we don't all_gather all predictions because of memory concerns and also because we don't want to make the assumption that the number of batches is evenly divisible by number of gpus. To use prediction "properly", the user currently has to implement the PredictionWriterCallback.

**Comment by gianscarpe** [comment]
Thanks @awaelchli for clarifying. In the end I solved it with
`def collect_results_gpu(result_part, size):
    rank, world_size = get_dist_info()
    # dump result part to tensor with pickle
    part_tensor = torch.tensor(
        bytearray(pickle.dumps(result_part)), dtype=torch.uint8, device='cuda')
    # gather all result part tensor shape
    shape_tensor = torch.tensor(part_tensor.shape, device='cuda')
    shape_list = [shape_tensor.clone() for _ in range(world_size)]
    dist.all_gather(shape_list, shape_tensor)
    # padding result part tensor to max length
    shape_max = torch.tensor(shape_list).max()
    part_send = torch.zeros(shape_max, dtype=torch.uint8, device='cuda')
    part_send[:shape_tensor[0]] = part_tensor
    part_recv_list = [
        part_tensor.new_zeros(shape_max) for _ in range(world_size)
    ]
    # gather all result part
    dist.all_gather(part_recv_list, part_send)

    if rank == 0:
        part_list = []
        for recv, shape in zip(part_recv_list, shape_list):
            part_list.append(
                pickle.loads(recv[:shape[0]].cpu().numpy().tobytes()))
        # sort the results
        ordered_results = []
        for res in zip(*part_list):
            ordered_results.extend(list(res))
        # the dataloader may pad some samples
        ordered_results = ordered_results[:size]
        return ordered_results`

**Comment by marcmk6** [comment]
> Thanks @awaelchli for clarifying. In the end I solved it with https://github.com/open-mmlab/mmdetection/blob/482f60fe55c364e50e4fc4b50893a25d8cc261b0/mmdet/apis/test.py#L160

Hi @gianscarpe, how did you integrate this into your code? I tried to override the hook [`on_prediction_epoch_end`](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-predict-epoch-end) but nothing is passed to it.
